File: src/common/py/data/entities/metadata/metadata_profile_container.py
from dataclasses import dataclass, field
from enum import StrEnum
from typing import List

from dataclasses_json import dataclass_json

from ..properties import ProfileMetadata, PropertyProfile


@dataclass_json
@dataclass(kw_only=True)
class MetadataProfileContainer:
    """
    A container that holds a metadata profile along with various descriptive information.

    Attributes:
        category: The overall category of the profile.
        role: The role of the profile within its category.
        profile: The actual profile data.
    """

    class Role(StrEnum):
        """
        The role of a profile.

        Attributes:
            DEFAULT: A default (fixed) profile that will always be used.
            OPTIONAL: An optional profile.
        """

        DEFAULT = "default"
        OPTIONAL = "optional"

    category: str
    role: Role

    profile: PropertyProfile = field(
        default_factory=lambda: PropertyProfile(
            metadata=ProfileMetadata(("", ""), "", ""), layout=[]
        )
    )


MetadataProfileContainerList = List[MetadataProfileContainer]

File: src/common/web/data/entities/metadata/MetadataProfileContainer.ts
import { PropertyProfile } from "../../../ui/components/propertyeditor/PropertyProfile";

/**
 * The role of a profile.
 */
export const enum MetadataProfileContainerRole {
    Template = "template",
    Global = "global"
}

/**
 * A container that holds a metadata profile along with various descriptive information.
 *
 * @param category - The overall category of the profile.
 * @param role - The role of the profile within its category.
 * @param profile - The actual profile data.
 */
export class MetadataProfileContainer {
    public readonly category: string;
    public readonly role: MetadataProfileContainerRole;

    public readonly profile: PropertyProfile;

    public constructor(category: string, role: MetadataProfileContainerRole, profile: PropertyProfile) {
        this.category = category;
        this.role = role;

        this.profile = profile;
    }
}

/**
 * A list of metadata profile containers.
 */
export type MetadataProfileContainerList = MetadataProfileContainer[];
